### DISCUSSION TITLE
PATCH a user via REST API?

### DISCUSSION THREAD

**Discussion by levelonestl** [drakkan/sftpgo#1156] [Q&A]
I have an event rule/action setup so that when a user is added/updated it will automatically create several directories automatically.  The issue I have is that when the list of directories that needs to be created is updated there is no way to automatically apply these new directories to existing users.  My thought was to maybe update some benign field on the user record via the REST API but it appears (unless I'm reading it incorrectly) that I would need to PUT the entire user record.  I was wondering if there was something else I could "touch" on a user that would fire that user update event or if there is some kind of PATCH functionality where I would just need to update a single field?

**Comment by drakkan** [comment]
Hello,

PATCH is not currently supported. You could add a scheduled rule that checks for the existence of the directories and, if it fails, create them. For example an action like this

![action](https://user-images.githubusercontent.com/553263/213259970-9bfd4a17-1267-4475-a7ee-a83279dbe398.png)

and a new scheduled rule with the following actions

![rule actions](https://user-images.githubusercontent.com/553263/213260040-5f9423ab-b709-4a89-a2cc-f0d9abffd03e.png)

Would this solve your issue?

EDIT: you can simply schedule your create directories action. The create directory action is a NOOP if the directory already exist

**Reply by levelonestl** [reply]
Would that still work if all my users are configured to use cloud storage (inherited from group membership)?  I could have upwards of 600 directories (users x directories) needing to be created and if I do it on a schedule even hourly it could be generating a half a million bucket actions per month.   I'm not allowing users to login via the web.  I think I might instead just poke the 2fa/disable url for each user every time I need to add a new directory instead (which is rare).

**Reply by drakkan** [reply]
> Would that still work if all my users are configured to use cloud storage (inherited from group membership)? 

yes it will work

> I could have upwards of 600 directories (users x directories) needing to be created and if I do it on a schedule even hourly it could be generating a half a million bucket actions per month. I'm not allowing users to login via the web. I think I might instead just poke the 2fa/disable url for each user every time I need to add a new directory instead (which is rare).

2fa/disable is a workaround it you don't allow web login

**Reply by drakkan** [reply]
> Would that still work if all my users are configured to use cloud storage (inherited from group membership)? I could have upwards of 600 directories (users x directories) needing to be created and if I do it on a schedule even hourly it could be generating a half a million bucket actions per month. I'm not allowing users to login via the web. I think I might instead just poke the 2fa/disable url for each user every time I need to add a new directory instead (which is rare).

I'm evaluating various possibilities to simplify this use case. Probably a good compromise is to allow you to define rules with inactive status and add the ability to run them on demand. This way you can just run the rule manually from the WebAdmin or REST API instead of updating all your old users. What do you think about?

**Reply by levelonestl** [reply]
Let me give you a better overview of what I'm doing.  I have different groups of users that need different directory structures pre-created upon user add/update.  The only solution I could find was to create 2 admin accounts, 2 provider event rules, and 2 corresponding event actions.  When I want to interact with one type (group) of user I use the corresponding admin account.  The groups that these users belong to define their storage type (google cloud storage) and have ACLs setup to only allow downloading and uploading to various directories (the ones created by the event actions).  I've not used the schedule type rules but it appears that I could create one that filters based on user groups correct? If it could then trigger the corresponding event action that seems like it would work as well.  What you are suggesting is making it so I could start the event rule(s) manually and have it attempt to apply my corresponding directories?

**Reply by drakkan** [reply]
> Let me give you a better overview of what I'm doing. I have different groups of users that need different directory structures pre-created upon user add/update. The only solution I could find was to create 2 admin accounts, 2 provider event rules, and 2 corresponding event actions. When I want to interact with one type (group) of user I use the corresponding admin account.

This is correct if you want group membership to be enforced automatically, otherwise a single administrator account should be fine

> The groups that these users belong to define their storage type (google cloud storage) and have ACLs setup to only allow downloading and uploading to various directories (the ones created by the event actions). 

great, you understood perfectly what groups are for

> I've not used the schedule type rules but it appears that I could create one that filters based on user groups correct? If it could then trigger the corresponding event action that seems like it would work as well. What you are suggesting is making it so I could start the event rule(s) manually and have it attempt to apply my corresponding directories?

I'm thinking of adding a new trigger named "On demand" or something similar. As with the "Schedule" trigger, you will be able to execute the defined actions for all users, filtering by name, groups (and roles in v2.5). "On demand" rules will be executed manually from the WebAdmin or using the REST API.

If I understand your use case correctly, you will need to run at most 2 "On demand" rules instead of updating all your users to trigger your rules. You will be able to reuse your already defined actions to create directories for the "On demand" rules.

**Reply by levelonestl** [reply]
I think that would work.

**Comment by drakkan** [comment]
https://github.com/drakkan/sftpgo/commit/7b5bebc588f54589f867001418ab052c162e5fbd

it will be included in v2.5.0, any testing using the development version would be greatly appreciated, thanks
